Support ActiveRecord 7.0.x

mshibuya opened the pull request:
This PR bumps up the ActiveRecord dependency to 7.0.x, while making necessary changes to keep it working. Changes are separated into individual commits, with the reason for the change and the link to the corresponding commit in ActiveRecord.

Any suggestions would be appreciated. Thanks!

mshibuya commented on activerecord-spanner-adapter.gemspec:
```diff
@@ -25,7 +25,7 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency "google-cloud-spanner", "~> 2.10"
-  spec.add_runtime_dependency "activerecord", "~> 6.1.4"
+  spec.add_runtime_dependency "activerecord", "~> 7.0.0"
```
I've dropped support for ActiveRecord 6.1.x, but if you wish I think we can keep compatibility both with ActiveRecord 6.1.x and 7.0.x. If you prefer to do so, please let me know.

hengfengli replied:
I think we would like to keep compatibility both with ActiveRecord 6.1.x and 7.0.x. @olavloite What do you think?

dazuma replied:
Ideally we'd want to keep compatibility with the last two major versions. If there are breaking changes between them, it should still be possible via logic in the code that branches based on the ActiveRecord version.

hengfengli commented:
@olavloite Can you please take a review on this PR? Thanks.

mshibuya commented:
Changed to keep support for ActiveRecord 6.0 and 6.1. Confirmed locally that the unit test and the acceptance test on emulator passes with all combinations.

Also added 7.0 to the GitHub Actions matrix, though I haven't actually ran it yet... (awaiting approval)

olavloite commented:
> Changed to keep support for ActiveRecord 6.0 and 6.1. Confirmed locally that the unit test and the acceptance test on emulator passes with all combinations.
> 
> Also added 7.0 to the GitHub Actions matrix, though I haven't actually ran it yet... (awaiting approval)

Thanks for adding that. I've approved the test run.

mshibuya commented:
All CI checks are passing 🎉 
What else can I do to get this merged?

palladius commented:
Can we please expedite the reviewing and merging?

olavloite commented on Gemfile:
```diff
@@ -3,6 +3,7 @@ source "https://rubygems.org"
 # Specify your gem's dependencies in activerecord-spanner.gemspec
 gemspec
 
+gem "activerecord", ENV.fetch("AR_VERSION", "~> 7.0.0")
```
I think that this would mean that we should also bump the major version number of the ActiveRecord provider, as we are changing the (default) ActiveRecord version from 6.x to 7.x. I think it would be better to keep this at 6.1.4 as the default version. We can always bump the default version at a later moment.

mshibuya replied:
Sure, I will!

olavloite commented on lib/activerecord_spanner_adapter/base.rb:
```diff
@@ -117,6 +125,7 @@ def self._create_grpc_values_for_insert metadata, values
       serialized_values = []
       columns = []
       values.each_pair do |k, v|
+        v = unwrap_attribute v
```
Is this a change that is needed for 7.x, or was this an existing bug in the current implementation?

mshibuya replied:
This is to catch up with this ActiveRecord 7.x change.
https://github.com/rails/rails/commit/6ee96a8f42d6b13bffd46342248f447d9f289288
With this change merged into ActiveRecord 7.0, we will receive `ActiveModel::Attribute` instances for the return value of `#attributes_with_values` so they need to be unwrapped. In ActiveRecord 6.x we get the values itself, hence [the conditional](https://github.com/googleapis/ruby-spanner-activerecord/pull/170/commits/dd650979d4d83e3d096b82333c64ed077c28b875#diff-2b0d1b3234f56c21e27aa0c68126687695874ba8603f41bd571c7847397b8988R91-R95) here.

olavloite reviewed (commented):
This looks generally good to me. The only real question I have is regarding the (apparent) update of the default ActiveRecord version from 6.x to 7.x. If we change the default major version that is used by this provider, I feel that we should also bump the major version number of the provider. It might be better if we could keep the default version at 6.x, but still add support for 7.x for those who want to opt-in for that.

(Also waiting for all the integration tests in https://github.com/googleapis/ruby-spanner-activerecord/pull/171 to finish before approving)

mshibuya commented:
> If we change the default major version that is used by this provider, I feel that we should also bump the major version number of the provider.

About this point my feeling is not, because newly adding support for ActiveRecord 7.0 with keeping compatibility with 6.x can be considered as a feature addition in a backward-compatible way, which doesn't trigger a major version bump according to [Semver](https://semver.org/#spec-item-7).
But of course it's totally up to you maintainers, this is just my thought.

olavloite reviewed (approved):
LGTM. Thanks for the nice work.

The integration test failure that was noticed in #171 is not related to this change, and will be fixed by #173
